### ISSUE TITLE
False positives in unused_private_declaration

### ISSUE THREAD

**Issue by marcelofabri** [realm/SwiftLint#2809]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

Verified by adding the examples in `UnusedPrivateDeclarationRule`'s `nonTriggeringExamples`.

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? master
* Installation method used (Homebrew, CocoaPods, building from source, etc)? Source

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents.

No

* Which Xcode version are you using (check `xcode-select -p`)? Xcode 10.2.1
* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

I've reduce these from real code:

```swift
extension UIViewController {
    func presentPrompt(image: UIImage?, message: String) {
        prompt(image: image, message: message)
    }

    // this triggers
    private func prompt(image: UIImage? = nil,
                        title: String? = nil,
                        message: String? = nil,
                        details: String? = nil) {}
}



final class Client {
    func isConnected() -> Single<Bool> {
        return method { [sessionManager] in
            sessionManager.request()
        }
    }

    // this triggers
    private let sessionManager: SessionManager = {
        return SessionManager()
    }()
}

struct Model {
    private let codableStartsAt: Date
	
    // this triggers
    var startsAt: Date {
        return codableStartsAt
    }
}
```

**Comment by jpsim** [comment]
Odd, I'll need to look into this one. Does it still trigger if you use an explicit `self.` reference? `self.prompt(image: image, message: message)`

```swift
extension UIViewController {
    func presentPrompt(image: UIImage?, message: String) {
        prompt(image: image, message: message)
    }

    // this triggers
    private func prompt(image: UIImage? = nil,
                        title: String? = nil,
                        message: String? = nil,
                        details: String? = nil) {}
}
```

I think this is a "known" limitation in that capture lists aren't checked for usage. I think this should be straightforward to fix. In the meantime, a workaround could be to bind it explicitly: `[sessionManager = self.sessionManager]`

```swift
final class Client {
    func isConnected() -> Single<Bool> {
        return method { [sessionManager] in
            sessionManager.request()
        }
    }

    // this triggers
    private let sessionManager: SessionManager = {
        return SessionManager()
    }()
}
```

I don't follow what you're saying, I don't see `startsAt` used in your sample, so this is a legitimate violation.

```swift
struct Model {
    private let codableStartsAt: Date
	
    // this triggers
    var startsAt: Date {
        return codableStartsAt
    }
}
```

**Comment by marcelofabri** [comment]
> I don't follow what you're saying, I don't see startsAt used in your sample, so this is a legitimate violation.

`startsAt` is not `private`, shouldn't it only check `private` declarations? (In my real world example, `startsAt` is definitely used)

**Comment by jpsim** [comment]
I see. This passes the test:

```swift
import Foundation

struct Model {
    private let codableStartsAt: Date

    private var ↓startsAt: Date {
        return codableStartsAt
    }
}
```

Are you importing `Foundation` in that file? If not, what happens if you do?

**Comment by marcelofabri** [comment]
> Are you importing `Foundation` in that file? If not, what happens if you do?

I'm not. If I add it to the example it passes though 😲

**Comment by gobetti** [comment]
I was also using Xcode 10.2.1 when I tested SwiftLint's `unused_private_declaration` and alternatives like Periphery and [this simple Ruby script](https://github.com/PaulTaykalo/swift-scripts/blob/master/unused.rb). All of the 3 have triggered many false positives with private properties and methods so I'm wondering if this isn't a bug somewhere else.
